panic: index out of range in StripedLockSet.Lock when KubeletEnsureSecretPulledImages is enabled

### What happened?

Kubelet panics on startup when attempting to sync a pod, causing it to exit with code 2 (INVALIDARGUMENT) in my systemd unit . The panic occurs in the StripedLockSet implementation in the new PullManager introduced for the KubeletEnsureSecretPulledImages feature gate.

Kubelet version: 1.35.4

Feature gate: KubeletEnsureSecretPulledImages (BETA, default=true)

Panic:
```
I0509 20:48:26.079646 526586 scope.go:122] "RemoveContainer" containerID="04ffb75032616f8582756fe449346693be962bdbf4654adfe125911ae57344f5"
E0509 20:48:26.080758 526586 panic.go:115] "Observed a panic" panic="runtime error: index out of range [63] with length 31" panicGoValue="runtime.boundsError{x:63, y:31, signed:true, code:0x0}" stacktrace=<
goroutine 2326 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x298f068, 0xc00284d710}, {0x23a03c0, 0xc000c2fd58})
k8s.io/apimachinery/pkg/util/runtime/runtime.go:132 +0xbc
k8s.io/apimachinery/pkg/util/runtime.handleCrash({0x298f030, 0x3b9cb00}, {0x23a03c0, 0xc000c2fd58}, {0xc002bcf8a0, 0x0, 0x485932?})
k8s.io/apimachinery/pkg/util/runtime/runtime.go:107 +0x116
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0x2966100?})
k8s.io/apimachinery/pkg/util/runtime/runtime.go:64 +0x17b
panic({0x23a03c0?, 0xc000c2fd58?})
runtime/panic.go:783 +0x132
k8s.io/kubernetes/pkg/kubelet/images/pullmanager.(*StripedLockSet).Lock(0xc000806900, {0xc0030d24b0?, 0xc0030d9270?})
k8s.io/kubernetes/pkg/kubelet/images/pullmanager/locks.go:44 +0x68
k8s.io/kubernetes/pkg/kubelet/images/pullmanager.(*PullManager).MustAttemptImagePull.func2(0xc0002f8f00, {0xc0030d24b0, 0x47}, {0xc000c70300, 0x51}, 0xc002bcfbc0, 0xc002bcfb15)
k8s.io/kubernetes/pkg/kubelet/images/pullmanager/image_pull_manager.go:204 +0x6f
k8s.io/kubernetes/pkg/kubelet/images/pullmanager.(*PullManager).MustAttemptImagePull(0xc0002f8f00, {0x298f0d8, 0xc002bae6d0}, {0xc000c70300, 0x51}, {0xc0030d24b0, 0x47}, 0xc0030d9260)
k8s.io/kubernetes/pkg/kubelet/images/pullmanager/image_pull_manager.go:236 +0x154
k8s.io/kubernetes/pkg/kubelet/images.(*imageManager).EnsureImageExists(0xc000914a00, {0x298f0d8, 0xc002bae6d0}, 0xc001ac1f10, 0xc001279908, {0xc000c70300, 0x51}, {0xc0011d6c80, 0x1, 0x1}, ...)
k8s.io/kubernetes/pkg/kubelet/images/image_manager.go:253 +0x90b
k8s.io/kubernetes/pkg/kubelet/kuberuntime.(*kubeGenericRuntimeManager).startContainer(0xc000a6c820, {0x298f0d8, 0xc002bae6d0}, {0xc001e3e740, 0x40}, 0xc002b8c900, 0xc002bd0b30, 0xc001279908, 0xc000afe000, {0xc0011d6c80, ...}, ...)
k8s.io/kubernetes/pkg/kubelet/kuberuntime/kuberuntime_container.go:215 +0x37c
k8s.io/kubernetes/pkg/kubelet/kuberuntime.(*kubeGenericRuntimeManager).SyncPod.func2({0x298f0d8, 0xc002bae6d0}, {0x2588a6f, 0x9}, {0x2588a6f, 0x9}, 0xc002bd0b30)
k8s.io/kubernetes/pkg/kubelet/kuberuntime/kuberuntime_manager.go:1648 +0x92a
k8s.io/kubernetes/pkg/kubelet/kuberuntime.(*kubeGenericRuntimeManager).SyncPod(0xc000a6c820, {0x298f0d8, 0xc002bae6d0}, 0xc001279908, 0xc000afe000, {0xc0011d6c80, 0x1, 0x1}, 0xc000e7a180, 0x0)
k8s.io/kubernetes/pkg/kubelet/kuberuntime/kuberuntime_manager.go:1706 +0x3ca3
k8s.io/kubernetes/pkg/kubelet.(*Kubelet).SyncPod(0xc0004f5408, {0x298f0a0, 0xc001829180}, 0x2, 0xc001279908, 0x0, 0xc000afe000)
k8s.io/kubernetes/pkg/kubelet/kubelet.go:2155 +0x1ed2
k8s.io/kubernetes/pkg/kubelet.(*podWorkers).podWorkerLoop.func1(...)
k8s.io/kubernetes/pkg/kubelet/pod_workers.go:1309
k8s.io/kubernetes/pkg/kubelet.(*podWorkers).podWorkerLoop(0xc000b584d0, {0xc000cb3b00, 0x24}, 0xc00124b2d0)
k8s.io/kubernetes/pkg/kubelet/pod_workers.go:1314 +0x57e
k8s.io/kubernetes/pkg/kubelet.(*podWorkers).UpdatePod.func1()
k8s.io/kubernetes/pkg/kubelet/pod_workers.go:970 +0x118
created by k8s.io/kubernetes/pkg/kubelet.(*podWorkers).UpdatePod in goroutine 585
k8s.io/kubernetes/pkg/kubelet/pod_workers.go:965 +0x20c5

panic: runtime error: index out of range [63] with length 31 [recovered, repanicked]

goroutine 2326 [running]:
k8s.io/apimachinery/pkg/util/runtime.handleCrash({0x298f030, 0x3b9cb00}, {0x23a03c0, 0xc000c2fd58}, {0xc002bcf8a0, 0x0, 0x485932?})
k8s.io/apimachinery/pkg/util/runtime/runtime.go:114 +0x1a9
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0x2966100?})
k8s.io/apimachinery/pkg/util/runtime/runtime.go:64 +0x17b
panic({0x23a03c0?, 0xc000c2fd58?})
runtime/panic.go:783 +0x132
k8s.io/kubernetes/pkg/kubelet/images/pullmanager.(*StripedLockSet).Lock(0xc000806900, {0xc0030d24b0?, 0xc0030d9270?})
k8s.io/kubernetes/pkg/kubelet/images/pullmanager/locks.go:44 +0x68
k8s.io/kubernetes/pkg/kubelet/images/pullmanager.(*PullManager).MustAttemptImagePull.func2(0xc0002f8f00, {0xc0030d24b0, 0x47}, {0xc000c70300, 0x51}, 0xc002bcfbc0, 0xc002bcfb15)
k8s.io/kubernetes/pkg/kubelet/images/pullmanager/image_pull_manager.go:204 +0x6f
k8s.io/kubernetes/pkg/kubelet/images/pullmanager.(*PullManager).MustAttemptImagePull(0xc0002f8f00, {0x298f0d8, 0xc002bae6d0}, {0xc000c70300, 0x51}, {0xc0030d24b0, 0x47}, 0xc0030d9260)
k8s.io/kubernetes/pkg/kubelet/images/pullmanager/image_pull_manager.go:236 +0x154
k8s.io/kubernetes/pkg/kubelet/images.(*imageManager).EnsureImageExists(0xc000914a00, {0x298f0d8, 0xc002bae6d0}, 0xc001ac1f10, 0xc001279908, {0xc000c70300, 0x51}, {0xc0011d6c80, 0x1, 0x1}, ...)
k8s.io/kubernetes/pkg/kubelet/images/image_manager.go:253 +0x90b
k8s.io/kubernetes/pkg/kubelet/kuberuntime.(*kubeGenericRuntimeManager).startContainer(0xc000a6c820, {0x298f0d8, 0xc002bae6d0}, {0xc001e3e740, 0x40}, 0xc002b8c900, 0xc002bd0b30, 0xc001279908, 0xc000afe000, {0xc0011d6c80, ...}, ...)
k8s.io/kubernetes/pkg/kubelet/kuberuntime/kuberuntime_container.go:215 +0x37c
k8s.io/kubernetes/pkg/kubelet/kuberuntime.(*kubeGenericRuntimeManager).SyncPod.func2({0x298f0d8, 0xc002bae6d0}, {0x2588a6f, 0x9}, {0x2588a6f, 0x9}, 0xc002bd0b30)
k8s.io/kubernetes/pkg/kubelet/kuberuntime/kuberuntime_manager.go:1648 +0x92a
k8s.io/kubernetes/pkg/kubelet/kuberuntime.(*kubeGenericRuntimeManager).SyncPod(0xc000a6c820, {0x298f0d8, 0xc002bae6d0}, 0xc001279908, 0xc000afe000, {0xc0011d6c80, 0x1, 0x1}, 0xc000e7a180, 0x0)
k8s.io/kubernetes/pkg/kubelet/kuberuntime/kuberuntime_manager.go:1706 +0x3ca3
k8s.io/kubernetes/pkg/kubelet.(*Kubelet).SyncPod(0xc0004f5408, {0x298f0a0, 0xc001829180}, 0x2, 0xc001279908, 0x0, 0xc000afe000)
k8s.io/kubernetes/pkg/kubelet/kubelet.go:2155 +0x1ed2
k8s.io/kubernetes/pkg/kubelet.(*podWorkers).podWorkerLoop.func1(...)
k8s.io/kubernetes/pkg/kubelet/pod_workers.go:1309
k8s.io/kubernetes/pkg/kubelet.(*podWorkers).podWorkerLoop(0xc000b584d0, {0xc000cb3b00, 0x24}, 0xc00124b2d0)
k8s.io/kubernetes/pkg/kubelet/pod_workers.go:1314 +0x57e
k8s.io/kubernetes/pkg/kubelet.(*podWorkers).UpdatePod.func1()
k8s.io/kubernetes/pkg/kubelet/pod_workers.go:970 +0x118
created by k8s.io/kubernetes/pkg/kubelet.(*podWorkers).UpdatePod in goroutine 585
k8s.io/kubernetes/pkg/kubelet/pod_workers.go:965 +0x20c5

```
Impact:

Kubelet crashes on every start attempt when a pod triggers MustAttemptImagePull. The crash loop causes the node to never become Ready.

Workaround:

Disable the feature gate in kubelet-config.json:
```

"featureGates": {
    "KubeletEnsureSecretPulledImages": false
}
```

KubeletEnsureSecretPulledImages is BETA/default=true, by disabling it bypasses PullManager.MustAttemptImagePull entirely, seems to be falling back to the previous image pull path

### What did you expect to happen?

Kubelet not to crash with existing feature gates defaulted 

No panic. Lock stripe index should always be within bounds of the backing array.



### How can we reproduce it (as minimally and precisely as possible)?

Steps to reproduce:

Run kubelet 1.35.4 with KubenetEnsureSecretPulledImages enabled (default)
Schedule a pod to the node
Kubelet panics during EnsureImageExists → MustAttemptImagePull → StripedLockSet.Lock
Expected behavior: No panic. Lock stripe index should always be within bounds of the backing array.


### Anything else we need to know?

Kubelet config json (before disabling the feature gate)

```
{
  "kind": "KubeletConfiguration",
  "apiVersion": "kubelet.config.k8s.io/v1beta1",
  "address": "0.0.0.0",
  "authentication": {
    "anonymous": {
      "enabled": false
    },
    "webhook": {
      "cacheTTL": "2m0s",
      "enabled": true
    },
    "x509": {
      "clientCAFile": "/etc/kubernetes/pki/ca.crt"
    }
  },
  "authorization": {
    "mode": "Webhook",
    "webhook": {
      "cacheAuthorizedTTL": "5m0s",
      "cacheUnauthorizedTTL": "30s"
    }
  },
  "clusterDomain": "cluster.local",
  "hairpinMode": "hairpin-veth",
  "readOnlyPort": 0,
  "cgroupDriver": "systemd",
  "cgroupRoot": "/",
  "featureGates": {
    "RotateKubeletServerCertificate": true,
  },
  "protectKernelDefaults": true,
  "serializeImagePulls": false,
  "serverTLSBootstrap": true,
  "tlsCipherSuites": [
    "TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256",
    "TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256",
    "TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305",
    "TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384",
    "TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305",
    "TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384",
    "TLS_RSA_WITH_AES_256_GCM_SHA384",
    "TLS_RSA_WITH_AES_128_GCM_SHA256"
  ],
  "clusterDNS": [
    "169.254.20.10"
  ],
  "evictionHard": {
    "memory.available": "100Mi",
    "nodefs.available": "10%",
    "nodefs.inodesFree": "5%"
  },
  "kubeReserved": {
    "cpu": "110m",
    "ephemeral-storage": "1Gi",
    "memory": "2510Mi"
  },
  "maxPods": 205,
  "providerID": "aws:///us-east-1a/i-xxxxxxxxxxxxx",
  "enforceNodeAllocatable": [
    "pods",
    "kube-reserved"
  ],
  "systemReserved": {
    "cpu": "250m",
    "ephemeral-storage": "5Gi",
    "memory": "2Gi"
  },
  "podPidsLimit": 1000,
  "kubeReservedCgroup": "/kube.slice",
  "kubeletCgroups": "/kube.slice",
  "systemReservedCgroup": "/system.slice",
  "eventRecordQPS": 0,
  "staticPodPath": "/etc/kubernetes/manifests",
  "registryPullQPS": 25,
  "registryBurst": 150,
  "imageGCHighThresholdPercent": 85,
  "imageGCLowThresholdPercent": 80,
  "cgroupsPerQOS": true,
  "runtimeRequestTimeout": "30m0s",
  "cpuCFSQuota": false,
  "maxParallelImagePulls": 500
}

```

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: v1.35.0
Kustomize Version: v5.7.1
Server Version: v1.35.4-eks-40737a8
```

</details>


### Cloud provider

<details>
Amazon EKS
</details>


### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
NAME="Flatcar Container Linux by Kinvolk"
ID=flatcar
ID_LIKE=coreos
VERSION=4081.3.6
VERSION_ID=4081.3.6
BUILD_ID=2025-09-12-1631
SYSEXT_LEVEL=1.0
PRETTY_NAME="Flatcar Container Linux by Kinvolk 4081.3.6 (LTS 2024)"
ANSI_COLOR="38;5;75"
HOME_URL="https://flatcar.org/"
BUG_REPORT_URL="https://issues.flatcar.org"
FLATCAR_BOARD="amd64-usr"
$ uname -a
#Linux ip-x.x.x.xec2.internal 6.6.106-flatcar #1 SMP PREEMPT_DYNAMIC Fri Sep 12 16:05:08 -00 2025 x86_64 AMD EPYC 7R13 Processor AuthenticAMD GNU/Linux

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>


### Install tools

<details>

</details>


### Container runtime (CRI) and version (if applicable)

<details>
```
Client:
 Version:	v1.7.3
 OS/Arch:	linux/amd64
 Git commit:	0a464409d0178e16d3d2bed36222937ec3fc9c77
 buildctl:
  Version:

Server:
 containerd:
  Version:	v1.7.21
  GitCommit:	174e0d1785eeda18dc2beba45e1d5a188771636b
 runc:
  Version:	1.1.13
  GitCommit:	58aa9203c123022138b22cf96540c284876a7910
```
</details>


### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>
```
1.7.1 CNI Plugin

Cilium 1.17.13
```
</details>




/sig node

--- a/pkg/kubelet/images/pullmanager/locks.go
+++ b/pkg/kubelet/images/pullmanager/locks.go
@@ -21,6 +21,11 @@ import (
 	"sync"
 )
 
+const (
+	minStripedLockSetSize int32 = 1
+	maxStripedLockSetSize int32 = 31
+)
+
 // StripedLockSet allows context locking based on string keys, where each key
 // is mapped to a an index in a size-limited slice of locks.
 type StripedLockSet struct {
@@ -32,10 +37,10 @@ type StripedLockSet struct {
 // used for locking context based on string keys.
 // The size will be normalized to stay in the <1, 31> interval.
 func NewStripedLockSet(size int32) *StripedLockSet {
-	size = max(size, 1) // make sure we're at least at size 1
+	size = min(maxStripedLockSetSize, max(minStripedLockSetSize, size))
 
 	return &StripedLockSet{
-		locks: make([]sync.Mutex, min(31, size)),
+		locks: make([]sync.Mutex, size),
 		size:  size,
 	}
 }

--- a/pkg/kubelet/images/pullmanager/locks_test.go
+++ b/pkg/kubelet/images/pullmanager/locks_test.go
@@ -1,0 +1,82 @@
+/*
+Copyright The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pullmanager
+
+import (
+	"fmt"
+	"testing"
+)
+
+func TestStripedLockSetNormalizesSize(t *testing.T) {
+	tests := []struct {
+		name string
+		size int32
+		want int32
+	}{
+		{
+			name: "negative",
+			size: -1,
+			want: 1,
+		},
+		{
+			name: "zero",
+			size: 0,
+			want: 1,
+		},
+		{
+			name: "below max",
+			size: 10,
+			want: 10,
+		},
+		{
+			name: "above max",
+			size: 500,
+			want: 31,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			lockSet := NewStripedLockSet(tt.size)
+			if len(lockSet.locks) != int(tt.want) {
+				t.Fatalf("len(NewStripedLockSet(%d).locks) = %d, want %d", tt.size, len(lockSet.locks), tt.want)
+			}
+		})
+	}
+}
+
+func TestStripedLockSetLargeRequestedSizeDoesNotPanic(t *testing.T) {
+	const requestedSize int32 = 500
+	lockSet := NewStripedLockSet(requestedSize)
+	key := keyWithOriginalStripeOutsideAllocatedRange(t, requestedSize, len(lockSet.locks))
+
+	lockSet.Lock(key)
+	lockSet.Unlock(key)
+}
+
+func keyWithOriginalStripeOutsideAllocatedRange(t *testing.T, requestedSize int32, allocatedLocks int) string {
+	t.Helper()
+
+	for i := range 1000 {
+		key := fmt.Sprintf("image-ref-%d", i)
+		if keyToID(key, requestedSize) >= uint32(allocatedLocks) {
+			return key
+		}
+	}
+	t.Fatalf("failed to find key with stripe outside allocated range")
+	return ""
+}

